charms/kfp-profile-controller/tests/unit/test_operator.py

```diff
@@ -1,194 +1,223 @@
 import yaml
-from oci_image import MissingResourceError
 from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
 from ops.testing import Harness
 
-from charm import CheckFailedError, KfpProfileControllerOperator
+from charm import KfpProfileControllerOperator
+
+MOCK_OBJECT_STORAGE_DATA = {
+    "access-key": "access-key",
+    "secret-key": "secret-key",
+    "service": "service",
+    "namespace": "namespace",
+    "port": 1234,
+    "secure": True,
+}
+
+EXPECTED_ENVIRONMENT = {
+    "CONTROLLER_PORT": 80,
+    "DISABLE_ISTIO_SIDECAR": "false",
+    "KFP_DEFAULT_PIPELINE_ROOT": "",
+    "KFP_VERSION": "2.0.0-alpha.7",
+    "METADATA_GRPC_SERVICE_HOST": "mlmd.kubeflow",
+    "METADATA_GRPC_SERVICE_PORT": "8080",
+    "MINIO_ACCESS_KEY": "access-key",
+    "MINIO_HOST": "service",
+    "MINIO_NAMESPACE": "namespace",
+    "MINIO_PORT": 1234,
+    "MINIO_SECRET_KEY": "secret-key",
+    "minio-secret": '{"secret": {"name": ' '"kfp-profile-controller-minio-credentials"}}',
+}
 
-# TODO: Tests missing for dropped/reloaded relations
 
+@pytest.fixture
+def harness() -> Harness:
+    harness = Harness(KfpProfileControllerOperator)
+    return harness
+
+
+@pytest.fixture()
+def mocked_lightkube_client(mocker):
+    """Mocks the Lightkube Client in charm.py, returning a mock instead."""
+    mocked_lightkube_client = MagicMock()
+    mocker.patch("charm.lightkube.Client", return_value=mocked_lightkube_client)
+    yield mocked_lightkube_client
+
+
+# Helpers
+def add_data_to_sdi_relation(
+    harness: Harness,
+    rel_id: int,
+    other: str,
+    data: Optional[dict] = None,
+    supported_versions: str = "- v1",
+) -> None:
+    """Add data to an SDI-backed relation."""
+    if data is None:
+        data = {}
+
+    harness.update_relation_data(
+        rel_id,
+        other,
+        {"_supported_versions": supported_versions, "data": yaml.dump(data)},
+    )
+
+
+def add_sdi_relation_to_harness(
+    harness: Harness, relation_name: str, other_app: str = "other", data: Optional[dict] = None
+) -> dict:
+    """Relates a new app and unit to an sdi-formatted relation.
+
+    Args:
+        harness: the Harness to add a relation to
+        relation_name: the name of the relation
+        other_app: the name of the other app that is relating to our charm
+        data: (optional) the data added to this relation
+
+    Returns dict of:
+    * other (str): The name of the other app
+    * other_unit (str): The name of the other unit
+    * rel_id (int): The relation id
+    * data (dict): The relation data put to the relation
+    """
+    if data is None:
+        data = {}
+
+    other_unit = f"{other_app}/0"
+    rel_id = harness.add_relation(relation_name, other_app)
+
+    harness.add_relation_unit(rel_id, other_unit)
+
+    add_data_to_sdi_relation(harness, rel_id, other_app, data)
+
+    return {
+        "other_app": other_app,
+        "other_unit": other_unit,
+        "rel_id": rel_id,
+        "data": data,
+    }
```

Comment:
You can now import these from chisme!  `from charmed_kubeflow_chisme.testing import add_sdi_relation_to_harness`

But beware, if you use the return value from that `add_sdi_relation_to_harness` it was modified slightly.  It used to return a dict, but now it returns a data class that actually shows your attributes.

Reply:
afaict, you never use the return from `add_sdi_relation_to_harness` so it shouldn't matter